apps/dashr-ocean-optics/app.R

```diff
@@ -0,0 +1,580 @@
+
+  function(...) {
+    args <- list(...)
+
+    argsTime <- args[[length(args)]]
+    now <- as.numeric(Sys.time()) * 1000
+
+    disabled <- list(
+      'color' = colors[['accent']],
+      'backgroundColor' = colors[['background']],
+      'cursor' = 'not-allowed'
+    )
+
+    enabled <- list(
+      'color' = colors[['background']],
+      'backgroundColor' = colors[['accent']],
+      'cursor' = 'pointer'
+    )
+
+    # if the button was recently clicked (less than a second ago), then
+    # it's safe to say that the callback was triggered by the button; so
+    # we have to "disable" it
+    if (now - argsTime < 500 && argsTime > 0) {
+      return(disabled)
+    } else {
+      return(enabled)
+    }
+  }
+)
+
+
+# disable/enable controls
+app$callback(
+  output = list(id = "controls", property = "children"),
+  params = list(input(id = "power-button", property = "on")),
+
+  function(pwr_on) {
+    lapply(1:length(controls),
+           function(i) {
+             controls[[i]]$props$children[[2]]$props$disabled <- !pwr_on
+             controls[[i]]
+           })
+  }
+)
+
+
+# disable/enable intensity knob
+app$callback(
+  output = list(id = "light-intensity-knob", property = "disabled"),
+  params = list(
+    input(id = "power-button", property = "on"),
+    input(id = "light-source-input", property = "value")
+  ),
+
+  function(pwr, ls) {
+    return(!(pwr && ls != "NULL" && ls != ""))
+  }
+)
+
+
+# send user-selected options to spectrometer
+app$callback(
+  output = list(id = "submit-status", property = "children"),
+  params = c(list(input(id = "submit-button", "n_clicks")),
+             controlValuesState,
+             list(state(id = "power-button", "on"))),
+
+  function(...) {
+    args <- list(...)
+
+    # handle `NULL` args`
+    nullCheck <- sapply(args, class)
+    nullCheck == "list"
+    if (sum(nullCheck == "list") > 0 ) {
+      args[[which(nullCheck == "list")]] <- "None"
+    }
+
+    # power-button off case
+    if (!args[[length(args)]]) {
+      return (
+        list(
+          "Press the power button to the top-right of the app,
+  then press the \"update\" button above to apply your options to the spectrometer."
+        )
+      )
+    }
+
+    # if submit-button clicked
+    if (args[[1]] > 0) {
+      sumList <- list()
+      nextLine <- list(htmlBr(), htmlBr())
+
+      if ("l1" %in% args ) {
+        sumList <- list(
+          "The following parameters were not successfully updated:")
+        sumList <- c(sumList, nextLine)
+        sumList <- c(sumList, list("LIGHT SOURCE: Lamp not found."))
+        sumList <- c(sumList, nextLine, list(htmlHr(), htmlBr()))
+      }
+
+      sumList <-  c(sumList,
+                    list("The following parameters were successfully updated:"),
+                    nextLine)
+
+      # append updated controls to sumList
+      for (i in 1:length(controls)) {
+
+        if (args[[i + 1]] == "l1") {
```

Comment:
Instead of using `next` to modify control flow here, I would probably handle this by subsetting `controls` to operate only on the elements for which `args[[i + 1]] == "l1"` is `FALSE`.